src/mqtt_client.rs

```diff
@@ -230,46 +193,38 @@ where
             topic, data, properties
         );
 
-        let mut buffer: GenericArray<u8, T> = GenericArray::default();
+        let mut buffer: [u8; T] = [0; T];
         let packet = serialize::publish_message(&mut buffer, topic, data, properties)
             .map_err(|e| Error::Protocol(e))?;
         self.write(packet)
     }
 
-    fn socket_is_connected(&self) -> Result<bool, N::Error> {
-        let mut socket_ref = self.socket.borrow_mut();
-
-        if socket_ref.is_none() {
-            // Attempt to open a socket from the network stack.
-            socket_ref.replace(self.network_stack.open(Mode::NonBlocking)?);
+    fn socket_is_connected(&mut self) -> Result<bool, N::Error> {
+        if self.socket.is_none() {
+            let socket = self.network_stack.socket()?;
+            self.socket.replace(socket);
```

Comment:
I know this is existing logic, but I wouldn't expect the name socket_is_connected() to be a fallible function with side effects and  create the socket. We could return false if the socket is none, else call network_stack.is_connected(). And then remove the Result<> from the return signature.

Reply:
The `embedded-nal::TcpClientStack::is_connected()` defines a fallible return signature, so we always have be fallible here as well. However, you make a good point about allocating the socket here. I'll see if I can find a more logical place for it.